synapse/cortex.py

```diff
@@ -4496,6 +4496,81 @@ async def _addView(self, vdef):
         await self.feedBeholder('view:add', pack, gates=[iden])
         return pack
 
+    async def delViewWithLayer(self, iden):
+        '''
+        Delete a Cortex View and its write Layer by iden.
+
+        Note:
+            Any children of the View will have their parent view updated to
+            the deleted View's parent (if present). The deleted layer will also
+            be removed from any Views which contain it in their layer stack.
+        '''
+        view = self.views.get(iden)
+        if view is None:
+            raise s_exc.NoSuchView(mesg=f'No such view {iden=}', iden=iden)
+
+        if view.info.get('protected'):
+            mesg = f'Cannot delete view ({iden}) that has protected set.'
+            raise s_exc.CantDelView(mesg=mesg)
+
+        layriden = view.layers[0].iden
+        pareiden = None
+        if view.parent is not None:
+            pareiden = view.parent.iden
+
+        return await self._push('view:delwithlayer', iden, view.layers[0].iden, newparent=pareiden)
+
+    @s_nexus.Pusher.onPush('view:delwithlayer', passitem=True)
+    async def _delViewWithLayer(self, viewiden, layriden, nexsitem, newparent=None):
+
+        if viewiden == self.view.iden:
+            raise s_exc.SynErr(mesg='Cannot delete the main view')
+
+        if (view := self.views.get(viewiden)) is not None:
+
+            await self.hive.pop(('cortex', 'views', viewiden))
+            await view.delete()
+
+            self._calcViewsByLayer()
+            await self.feedBeholder('view:del', {'iden': viewiden}, gates=[viewiden])
+            await self.auth.delAuthGate(viewiden)
+
+        if newparent is not None:
+            newview = self.views.get(newparent)
+
+        for cview in self.views.values():
+            if cview.parent is not None and cview.parent.iden == viewiden:
+                if newparent is None:
+                    cview.parent = None
+                    await cview.info.pop('parent')
+                else:
+                    cview.parent = newview
+                    await cview.info.set('parent', newparent)
+
+        if (layr := self.layers.get(layriden)) is not None:
+            del self.layers[layriden]
+
+            for pdef in layr.layrinfo.get('pushs', {}).values():
+                await self.delActiveCoro(pdef.get('iden'))
+
+            for pdef in layr.layrinfo.get('pulls', {}).values():
+                await self.delActiveCoro(pdef.get('iden'))
+
+            await self.feedBeholder('layer:del', {'iden': layriden}, gates=[layriden])
+            await self.auth.delAuthGate(layriden)
+            self.dynitems.pop(layriden)
+
+            await self.hive.pop(('cortex', 'layers', layriden))
+
+            await layr.delete()
+
+            layr.deloffs = nexsitem[0]
+
+        for cview in self.views.values():
+            if layr in cview.layers:
```

Comment:
It seems like we should only be deleting this layer from other views if the other view is a child of the current view (the only being deleted)? What if I had manually constructed a view with this layer in it? It just disappears from under me? What if it was my current write layer?

Reply:
Yeah I suppose we should leave it alone if it is used outside the current stack.